pkg/suggest/rules.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package suggest
+
+import (
+	"errors"
+	"fmt"
+	"github.com/agext/levenshtein"
+	"regexp"
+	"strings"
+)
+
+type rule struct {
+	condition      conditionFunc
+	transformation transformFunc
+}
+
+// conditionFunc is a function that returns true if the rule should be applied to the error.
+type conditionFunc func(error) bool
+
+// transformFunc is a function that defines how the error should be transformed.
+type transformFunc func(error) error
+
+// ruleInterface represents a suggestion rule.
+type ruleInterface interface {
+	apply(error) error
+}
+
+// newRule creates a new ruleInterface instance.
+func newRule(condition conditionFunc, transform transformFunc) ruleInterface {
+	return &rule{
+		condition:      condition,
+		transformation: transform,
+	}
+}
+
+// apply executes the rule on the error.
+func (g *rule) apply(err error) error {
+	if g.condition(err) {
+		return g.transformation(err)
+	}
+	return err
+}
+
+// newStrReplaceRule creates a ruleInterface that finds and replaces a string in the error message
+func newStrReplaceRule(find, replace string) ruleInterface {
+	condition := func(e error) bool {
+		return strings.Contains(e.Error(), find)
+	}
+
+	transformation := func(e error) error {
+		replacedMessage := strings.ReplaceAll(e.Error(), find, replace)
+		return errors.New(replacedMessage)
+	}
+
+	return newRule(condition, transformation)
+}
+
+// newRegexRule creates a ruleInterface based on a regex pattern.
+func newRegexRule(pattern string, transform transformFunc) ruleInterface {
+	re := regexp.MustCompile(pattern)
```

Comment:
We should be careful calling `regex.MustCompile` with a dynamic input since it panics if the regex is invalid. Not sure how we are providing `pattern` here as it seems to be a private function but mentioning just in case

Reply:
Agree, the purpose of this PR is to improve the user experience regarding the errors. This will potentially throw a panic which is not very descriptive usually. Couldn't we use `Compile` and handle the error gracefully?

Reply:
Thanks for the feedback, I will address it immediately